File: src/main/java/org/lanternpowered/server/inventory/PeekedTransactionResult.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.inventory;

import com.google.common.base.MoreObjects;
import com.google.common.collect.ImmutableList;
import org.spongepowered.api.item.inventory.transaction.InventoryTransactionResult;
import org.spongepowered.api.item.inventory.transaction.SlotTransaction;

import java.util.List;

public abstract class PeekedTransactionResult {

    private final List<SlotTransaction> transactions;

    /**
     * Constructs a new {@link PeekedTransactionResult}
     * from the given {@link SlotTransaction}s.
     *
     * @param transactions The slot transactions
     */
    protected PeekedTransactionResult(List<SlotTransaction> transactions) {
        this.transactions = ImmutableList.copyOf(transactions);
    }

    /**
     * Gets all the {@link SlotTransaction}s that occur
     * when the transaction result is applied.
     *
     * @return The slot transactions
     */
    public List<SlotTransaction> getTransactions() {
        return this.transactions;
    }

    /**
     * Accepts all the changes that are applied by this result,
     * only {@link SlotTransaction}s that aren't valid will be
     * ignored.
     */
    public void accept() {
        this.transactions.forEach(transaction -> {
            if (transaction.isValid()) {
                transaction.getSlot().set(transaction.getFinal().createStack());
            }
        });
    }

    /**
     * Gets whether this {@link PeekedTransactionResult} is empty.
     *
     * @return Is empty
     */
    public boolean isEmpty() {
        return this.transactions.isEmpty();
    }

    /**
     * Converts this {@link PeekedTransactionResult} into a
     * {@link InventoryTransactionResult}.
     *
     * @return The inventory transaction result
     */
    public InventoryTransactionResult asInventoryTransaction() {
        return asInventoryTransactionBuilder().build();
    }

    protected InventoryTransactionResult.Builder asInventoryTransactionBuilder() {
        return InventoryTransactionResult.builder()
                .type(InventoryTransactionResult.Type.SUCCESS) // Default to success
                .transaction(this.transactions);
    }

    protected MoreObjects.ToStringHelper toStringHelper() {
        return MoreObjects.toStringHelper(this)
                .add("transactions", this.transactions);
    }

    @Override
    public String toString() {
        return toStringHelper().toString();
    }
}

File: src/main/kotlin/org/lanternpowered/api/item/inventory/PeekedTransactionResult.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.api.item.inventory

import org.lanternpowered.api.item.inventory.transaction.SlotTransaction

/**
 * Represents a result of a peek operation.
 */
interface PeekedTransactionResult {

    /**
     * Gets all the [SlotTransaction]s that occur when the
     * transaction result is applied.
     *
     * @return The slot transactions
     */
    val transactions: List<SlotTransaction>

    /**
     * Gets whether this [PeekedTransactionResult] is empty.
     *
     * @return Is empty
     */
    val isEmpty: Boolean

    /**
     * Accepts all the changes that are applied by this result,
     * only [SlotTransaction]s that aren't valid will be ignored.
     *
     * Returns the [InventoryTransactionResult] that happened by
     * accepting this.
     */
    fun accept(): InventoryTransactionResult

    /**
     * Accepts all the changes that are applied by this result,
     * only [SlotTransaction]s that aren't valid will be ignored.
     *
     * This is the fast variant of [accept], use this if you don't care
     * about the transaction result.
     */
    fun acceptFast()
}
